### PR TITLE
fruit test

### PR DESCRIPTION
**Pull request by Daedrus1**

### CODE REVIEW

**Comment by JJJazl** [comment]
Please, do commits with informative messages

File: src/test/java/core/basesyntax/model/FruitTransactionTest.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+class FruitTransactionTest {
+    @Test
+    void testConstructorAndGetters() {
+        FruitTransaction.Operation operation = FruitTransaction.Operation.BALANCE;
+        String fruit = "apple";
+        int quantity = 100;
```
Comment (JJJazl) [inline src/test/java/core/basesyntax/model/FruitTransactionTest.java:13]: Move to the constants

File: src/test/java/core/basesyntax/io/FileWriterImplTest.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    private FileWriterImpl fileWriter;
+    private final String filePath = "testFile.txt";
+
+    @BeforeEach
+    void setUp() {
+        fileWriter = new FileWriterImpl();
+    }
+
+    @Test
+    void testWriteDataSuccessfully() throws IOException {
```
Comment (JJJazl) [inline src/test/java/core/basesyntax/io/FileWriterImplTest.java:26]: Please, follow the method name convention
For this task use such convention: <methodUnderTest>_<state>_<expectedBehavior>

File: src/test/java/core/basesyntax/io/FileWriterImplTest.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
+            String line = reader.readLine();
+            assertEquals(data, line,
+                    "The content written to the file should match the expected data");
+        }
+    }
+
+    @Test
+    void testFileCreation() {
```
Comment (JJJazl) [inline src/test/java/core/basesyntax/io/FileWriterImplTest.java:38]: Please, follow the method name convention

File: src/test/java/core/basesyntax/io/FileWriterImplTest.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        String data = "Checking file creation";
+        fileWriter.write(data, filePath);
+
+        File file = new File(filePath);
+        assertTrue(file.exists(), "File should be created when data is written to it");
+    }
+
+    @Test
+    void testIoExceptionHandling() {
```
Comment (JJJazl) [inline src/test/java/core/basesyntax/io/FileWriterImplTest.java:47]: Please, follow the method name convention

File: src/test/java/core/basesyntax/model/FruitTransactionTest.java [hunk 5]
```diff
@@ -0,0 +1,96 @@
+package core.basesyntax.model;
+
+import static org.junit.Assert.assertThrows;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+import org.junit.jupiter.api.Test;
+
+class FruitTransactionTest {
+    @Test
+    void testConstructorAndGetters() {
```
Comment (JJJazl) [inline src/test/java/core/basesyntax/model/FruitTransactionTest.java:10]: 1) Please, follow the method name convention
2) I think we don't need to test getters/setters separately because it will be tested in other methods anyway + they don't contain any logic

File: src/test/java/core/basesyntax/model/FruitTransactionTest.java [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+        FruitTransaction transaction = new FruitTransaction(operation, fruit, quantity);
+
+        assertEquals(operation, transaction.getOperation());
+        assertEquals(fruit, transaction.getFruit());
+        assertEquals(quantity, transaction.getQuantity());
+    }
+
+    @Test
+    void testSetAndGetOperation() {
```
Comment (JJJazl) [inline src/test/java/core/basesyntax/model/FruitTransactionTest.java:23]: 1) Please, follow the method name convention
2) I think we don't need to test getters/setters separately because it will be tested in other methods anyway + they don't contain any logic

File: src/test/java/core/basesyntax/model/FruitTransactionTest.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+        fruitTransaction.setOperation(FruitTransaction.Operation.BALANCE);
+        assertEquals(fruitTransaction.getOperation(), FruitTransaction.Operation.BALANCE);
+
+        fruitTransaction.setOperation(FruitTransaction.Operation.SUPPLY);
+        assertEquals(fruitTransaction.getOperation(), FruitTransaction.Operation.SUPPLY);
+
+    }
+
+    @Test
+    void testSetAndGetFruit() {
```
Comment (JJJazl) [inline src/test/java/core/basesyntax/model/FruitTransactionTest.java:34]: 1) Please, follow the method name convention
2) I think we don't need to test getters/setters separately because it will be tested in other methods anyway + they don't contain any logic

File: src/test/java/core/basesyntax/model/FruitTransactionTest.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+
+        transaction.setQuantity(100);
+        assertEquals(100, transaction.getQuantity());
+
+        transaction.setQuantity(200);
+        assertEquals(200, transaction.getQuantity());
+    }
+
+    @Test
+    void testGetByCodeForValidCodes() {
```
Comment (JJJazl) [inline src/test/java/core/basesyntax/model/FruitTransactionTest.java:55]: Please, follow the method name convention

File: src/test/java/core/basesyntax/model/FruitTransactionTest.java [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+                FruitTransaction.Operation.getByCode("p"),
+                "Code 'p' should return Operation.PURCHASE");
+
+        assertEquals(FruitTransaction.Operation.RETURN,
+                FruitTransaction.Operation.getByCode("r"),
+                "Code 'r' should return Operation.RETURN");
+    }
+
+    @Test
+    void testGetByCodeForInvalidCode() {
```
Comment (JJJazl) [inline src/test/java/core/basesyntax/model/FruitTransactionTest.java:74]: 1) Please, follow the method name convention
2) If you want to test different cases with method `getByCode()` you should have separate method for each scenario

File: src/test/java/core/basesyntax/service/DataConverterImplTest.java [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+class DataConverterImplTest {
+    private final DataConverterImpl dataConverter = new DataConverterImpl();
+
+    @Test
+    void testConvertToTransactionWithValidData() {
+        List<String> inputData = new ArrayList<>();
+        inputData.add("b,apple,100");
+        inputData.add("s,banana,50");
+        inputData.add("p,orange,25");
+        inputData.add("r,apple,10");
```
Comment (JJJazl) [inline src/test/java/core/basesyntax/service/DataConverterImplTest.java:21]: Move to the constatns and reuse in test methods

File: src/test/java/core/basesyntax/service/OperationStrategyImplTest.java [hunk 11]
```diff
@@ -0,0 +1,76 @@
...
+import core.basesyntax.service.operation.OperationHandler;
+import core.basesyntax.service.operation.PurchaseOperation;
+import core.basesyntax.service.operation.ReturnOperation;
+import core.basesyntax.service.operation.SupplyOperation;
+import java.util.HashMap;
+import java.util.Map;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class OperationStrategyImplTest {
```
Comment (JJJazl) [inline src/test/java/core/basesyntax/service/OperationStrategyImplTest.java:18]: Please, follow the method name convention

File: src/test/java/core/basesyntax/service/ReportGeneratorImplTest.java [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+
+import static org.junit.Assert.assertEquals;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import org.junit.jupiter.api.Test;
+
+class ReportGeneratorImplTest {
+    @Test
+    void testGetReportWithMultipleFruits() {
```
Comment (JJJazl) [inline src/test/java/core/basesyntax/service/ReportGeneratorImplTest.java:11]: Please, follow the method name convention

File: src/test/java/core/basesyntax/service/ShopServiceImplTest.java [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+import core.basesyntax.service.operation.PurchaseOperation;
+import core.basesyntax.service.operation.ReturnOperation;
+import core.basesyntax.service.operation.SupplyOperation;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+class ShopServiceImplTest {
```
Comment (JJJazl) [inline src/test/java/core/basesyntax/service/ShopServiceImplTest.java:17]: Please, follow the method name convention

**Review by JJJazl — CHANGES_REQUESTED** [review]

File: src/test/java/core/basesyntax/io/FileWriterImplTest.java (hunk 2 above)
Comment (Daedrus1) [reply src/test/java/core/basesyntax/io/FileWriterImplTest.java:26]: what rules to follow ?

File: src/main/java/core/basesyntax/service/DataConverter.java [hunk 14]
```diff
@@ -0,0 +1,9 @@
+package core.basesyntax.service;
+
+import core.basesyntax.model.FruitTransaction;
+import java.util.List;
+
+public interface DataConverter {
+    List<FruitTransaction> convertToTransaction(List<String> inputData);
+}
+
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/service/DataConverter.java:9]:
```suggestion
```

File: src/main/java/core/basesyntax/service/operation/ShopServiceImpl.java [hunk 15]
```diff
@@ -0,0 +1,12 @@
+package core.basesyntax.service.operation;
+
+import core.basesyntax.model.FruitTransaction;
+import core.basesyntax.service.ShopService;
+import java.util.List;
+
+public class ShopServiceImpl implements ShopService {
+    @Override
+    public void process(List<FruitTransaction> transactions) {
+
```
Comment (liliia-ponomarenko) [inline src/main/java/core/basesyntax/service/operation/ShopServiceImpl.java:10]: Let's realize this method. Or remove it if you don’t  use it

File: src/test/java/core/basesyntax/io/FileReaderImplTest.java [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Test
+    public void shouldReturnEmptyListWhenFileIsEmpty() throws IOException {
+        Files.createFile(Path.of(EMPTY_FILE_PATH));
+
+        List<String> lines = fileReader.read(EMPTY_FILE_PATH);
+        assertTrue(lines.isEmpty(), "Expected empty list for empty file");
+
+        // Удаляем файл после теста
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/io/FileReaderImplTest.java:38]: Let's remove redundant comment

File: src/test/java/core/basesyntax/io/FileReaderImplTest.java [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+    }
+
+    @Test
+    public void read_existingFile_ok() {
+        List<String> lines = fileReader.read(EXISTING_FILE_PATH);
+        assertEquals(6, lines.size());
+    }
+
+    @Test
+    public void shouldReturnEmptyListWhenFileIsEmpty() throws IOException {
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/io/FileReaderImplTest.java:32]: use unify format of the method names (check all test cases): 

Remember about the naming of test methods.
There are a lot of ways to name your test methods. The main point is that they should have informative names and be consistent along with other developers in your team. For this task use such convention: `<methodUnderTest>_<state>_<expectedBehavior>`. For example, if we are testing the method `register` with a `null` user's age the test method name should be `register_nullAge_notOk`. notOk is because the test expects the register method to throw an exception.

File: src/test/java/core/basesyntax/io/FileWriterImplTest.java [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+
+    @Test
+    void shouldWriteDataCorrectly_whenFilePathIsValid() throws IOException {
+        String data = "Hello, FileWriter!";
+        fileWriter.write(data, filePath);
+
+        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
+            String line = reader.readLine();
+            assertEquals(data, line,
+                    "The content written to the file should match the expected data");
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/io/FileWriterImplTest.java:33]: move check after try block. Also, it is better to use try-catch and move it in a private method

File: src/test/java/core/basesyntax/io/FileWriterImplTest.java [hunk 19]
```diff
@@ -0,0 +1,66 @@
...
+        }
+    }
+
+    @Test
+    void shouldCreateFile_whenDataIsWritten() {
+        String data = "Checking file creation";
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/io/FileWriterImplTest.java:39]: better test it with report content

File: src/test/java/core/basesyntax/model/FruitTransactionTest.java [hunk 20]
```diff
@@ -0,0 +1,81 @@
...
+
+    }
+
+    @Test
+    void shouldReturnCorrectFruit_whenSetFruitIsCalled() {
+        FruitTransaction fruitTransaction = new FruitTransaction(null, "", 0);
+        fruitTransaction.setFruit("banana");
+        assertEquals(fruitTransaction.getFruit(), "banana");
+
+        fruitTransaction.setFruit("apple");
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/model/FruitTransactionTest.java:27]: Don't test getters and setters of the model

File: src/test/java/core/basesyntax/model/FruitTransactionTest.java [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+
+        transaction.setQuantity(100);
+        assertEquals(100, transaction.getQuantity());
+
+        transaction.setQuantity(200);
+        assertEquals(200, transaction.getQuantity());
+    }
+
+    @Test
+    void shouldReturnCorrectOperation_whenValidCodeIsProvided() {
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/model/FruitTransactionTest.java:43]: better move it to different test cases

File: src/test/java/core/basesyntax/model/FruitTransactionTest.java [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+                FruitTransaction.Operation.getByCode("p"),
+                "Code 'p' should return Operation.PURCHASE");
+
+        assertEquals(FruitTransaction.Operation.RETURN,
+                FruitTransaction.Operation.getByCode("r"),
+                "Code 'r' should return Operation.RETURN");
+    }
+
+    @Test
+    void shouldThrowException_whenInvalidOrNullCodeIsProvided() {
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/model/FruitTransactionTest.java:62]: the same

File: src/test/java/core/basesyntax/service/DataConverterImplTest.java [hunk 23]
```diff
@@ -0,0 +1,63 @@
...
+        assertTrue(exception.getMessage().contains("Invalid input format"),
+                "Exception message should indicate an invalid input format");
+    }
+
+    @Test
+    void shouldThrowException_whenQuantityIsInvalid() {
+        List<String> inputData = new ArrayList<>();
+        inputData.add(INVALID_QUANTITY_APPLE);
+
+        NumberFormatException exception = assertThrows(NumberFormatException.class, () -> {
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/DataConverterImplTest.java:56]: you don't handle this exception in your solution. So add this logic or remove this test case

File: src/test/java/core/basesyntax/service/OperationStrategyImplTest.java [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+        operationHandlerMap.put(FruitTransaction.Operation.RETURN, returnHandler);
+
+        operationStrategy = new OperationStrategyImpl(operationHandlerMap);
+    }
+
+    @Test
+    void shouldReturnBalanceHandler_whenOperationIsBalance() {
+        OperationHandler handler = operationStrategy.get(FruitTransaction.Operation.BALANCE);
+        assertNotNull(handler, "Handler for BALANCE should not be null");
+        assertTrue(handler instanceof BalanceOperation,
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/OperationStrategyImplTest.java:43]: What is the difference between getClass() and instanceof? What will be better to use in this case?

File: src/test/java/core/basesyntax/service/ReportGeneratorImplTest.java [hunk 25]
```diff
@@ -0,0 +1,24 @@
...
+        Map<String, Integer> testMap = new LinkedHashMap<>();
+        testMap.put("banana", 100);
+        testMap.put("apple", 200);
+
+        ReportGenerator reportGenerator = new ReportGeneratorImpl(testMap);
+        String expectedReport = "fruit,quantity" + System.lineSeparator()
+                + "banana, 100" + System.lineSeparator()
+                + "apple, 200" + System.lineSeparator();
+        assertEquals(expectedReport, reportGenerator.getReport());
+
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/ReportGeneratorImplTest.java:21]:
```suggestion
```

File: src/test/java/core/basesyntax/service/ReportGeneratorImplTest.java [hunk 26]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/ReportGeneratorImplTest.java:23]:
```suggestion
```

File: src/test/java/core/basesyntax/service/ReportGeneratorImplTest.java [hunk 27]
```diff
@@ -0,0 +1,24 @@
+package core.basesyntax.service;
+
+import static org.junit.Assert.assertEquals;
+
+import java.util.LinkedHashMap;
+import java.util.Map;
+import org.junit.jupiter.api.Test;
+
+class ReportGeneratorImplTest {
```
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/ReportGeneratorImplTest.java:9]: Let’s  add test with empty map

File: src/test/java/core/basesyntax/service/ShopServiceImplTest.java (hunk 13 above)
Comment (liliia-ponomarenko) [inline src/test/java/core/basesyntax/service/ShopServiceImplTest.java:17]: Let's add some invalid cases. For example when there are not enough products to buy (also add the corresponding check to your solution if it is missing)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution ;)

File: src/test/java/core/basesyntax/io/FileReaderImplTest.java (hunk 17 above)
Comment (Daedrus1) [reply src/test/java/core/basesyntax/io/FileReaderImplTest.java:32]: I use the template should...when...

File: src/test/java/core/basesyntax/service/OperationStrategyImplTest.java (hunk 24 above)
Comment (Daedrus1) [reply src/test/java/core/basesyntax/service/OperationStrategyImplTest.java:43]: better to use instanceof, since it allows more flexibility in dealing with possible class descendants.

**Review by okuzan — APPROVED** [review]
